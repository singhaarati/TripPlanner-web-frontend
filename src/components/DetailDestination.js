import React, { useEffect, useState } from 'react'
import { Button, Card, Col, ListGroup, Modal, Row } from 'react-bootstrap'
import { Link, useNavigate, useParams } from 'react-router-dom'
import destinationService from '../services/destinationService'
import Footer from './Dashboard/Footer'
import TopBar from './Dashboard/TopBar'

export default function DetailDestination() {
    const [destination, setDestination] = useState([])
    const [reviews, setReviews] = useState([])
    const [loading, setLoading] = useState(true);
    const [newReviewText, setNewReviewText] = useState('');
    const [showEditModal, setShowEditModal] = useState(false);
    const [selectedReviewId, setSelectedReviewId] = useState('');
    const [editedText, setEditedText] = useState('');
    const { id } = useParams()
    const navigate = useNavigate();

    useEffect(() => {
        destinationService.getDestinationById(id)
            .then(res => {
                console.log(res.data.data)
                setDestination(res.data.data)
                setLoading(false)
            }).catch((err) => window.alert(err.response.data.error))
    }, [])

    useEffect(() => {
        destinationService.getAllReviews(id)
            .then(res => {
                console.log(res.data)
                setReviews(res.data)
            }).catch(err => window.alert(err.response.data.error))
    }, [])

    const handleDelete = (destinationId) => {
        destinationService.deleteDestination(destinationId)
            .then((res) => {
                setDestination((prevDestination) => {
                    return prevDestination.filter((dest) => dest.id !== destinationId);
                });
                navigate('/');
            }).catch((err) => console.log(err))
    }

    const handleAddReview = () => {
        if (!newReviewText.trim()) {
            window.alert("Review cannot be empty.");
            return;
        }
        destinationService
            .createReview(id, { text: newReviewText })
            .then((res) => {
                const createdReview = res.data.data;
                setReviews((prevReviews) => [createdReview, ...prevReviews]);
                setNewReviewText('');
                destinationService.getAllReviews(id)
                    .then(res => {
                        setReviews(res.data);
                    })
                    .catch(err => window.alert(err.response.data.error));
            })
            .catch((err) => console.log(err));
    };

    const handleDeleteReview = (destinationId, reviewId) => {
        destinationService.deleteReviewById(destinationId, reviewId)
            .then((res) => {
                setReviews((prevReviews) => {
                    return prevReviews.filter((review) => review.id !== reviewId);
                });
            })
            .catch((err) => {
                console.log(err);
            });
    }

    const handleEditReview = (reviewId, updatedText) => {
        destinationService.updateReviewById(destination.id, reviewId, { text: updatedText })
            .then((res) => {
                setReviews((prevReviews) => {
                    return prevReviews.map((review) => {
                        if (review.id === reviewId) {
                            return { ...review, text: updatedText };
                        }
                        return review;
                    });
                });
                setShowEditModal(false);
            })
            .catch((err) => {
                console.log(err);
            });
    };

    const handleOpenEditModal = (reviewId, reviewText) => {
        setSelectedReviewId(reviewId);
        setEditedText(reviewText);
        setShowEditModal(true);
    };

    return (
        <div>
            <TopBar />

            {loading ? (
                <div>Loading...</div>
            ) : (
                <ListGroup className='mb-3'>
                    <ListGroup.Item
                        key={destination.id}
                        className='d-flex justify-content-between align-items-start'>
                        <div className='ms-2 me-auto'>
                            <Link to={`/destination/${destination.id}`}>
                                <img
                                    src={`http://localhost:3007/uploads/${destination.picture}`}
                                    alt={destination.name}
                                    className="img-fluid rounded"
                                    style={{ maxHeight: '100%', width: '100%', objectFit: 'cover' }}
                                />
                            </Link>

                            <h2 style={{ marginBottom: '10px', color: '#333' }}>{destination.name}</h2>
                            <p style={{ marginBottom: '5px', color: '#666' }}>{destination.location}</p>
                            <p style={{ marginBottom: '20px', color: '#666' }}>{destination.price}</p>

                            {reviews.length > 0 ? (
                                <Row>
                                    {reviews.map((review, index) => (
                                        <Col md={6} key={index}>
                                            <Card className="mb-3">
                                                <Card.Body>
                                                    {review && (
                                                        <>
                                                            <Card.Title>{review.text}</Card.Title>
                                                            <Card.Text>{review.user}</Card.Text>

                                                            {/* Review Edit Button */}
                                                            <Button
                                                                variant="warning"
                                                                style={{ marginRight: '5px' }}
                                                                onClick={() => handleOpenEditModal(review.id, review.text)}
                                                            >
                                                                Edit
                                                            </Button>

                                                            {/*Review Delete Button */}
                                                            <Button
                                                                variant="danger"
                                                                onClick={() => handleDeleteReview(destination.id, review.id)}
                                                            >
                                                                Delete
                                                            </Button>
                                                        </>
                                                    )}
                                                </Card.Body>
                                            </Card>
                                        </Col>
                                    ))}
                                </Row>
                            ) : (
                                <p>No reviews available for this destination.</p>
                            )}

                            {/* New text box for creating review */}
                            <div className="mb-3">
                                <div className="input-group">
                                    <input
                                        type="text"
                                        value={newReviewText}
                                        onChange={(e) => setNewReviewText(e.target.value)}
                                        placeholder="Write a review..."
                                        className="form-control"
                                    />
                                    <button className="btn btn-primary" onClick={handleAddReview}>
                                        Add Review
                                    </button>
                                </div>
                            </div>

                            {/* Modal for editing reviews */}
                            <Modal show={showEditModal} onHide={() => setShowEditModal(false)}>
                                <Modal.Header closeButton>
                                    <Modal.Title>Edit Review</Modal.Title>
                                </Modal.Header>
                                <Modal.Body>
                                    <textarea
                                        value={editedText}
                                        onChange={(e) => setEditedText(e.target.value)}
                                        placeholder="Enter your review"
                                        rows="5"
                                        style={{ width: '100%' }}
                                    />
                                </Modal.Body>
                                <Modal.Footer>
                                    <Button variant="secondary" onClick={() => setShowEditModal(false)}>
                                        Cancel
                                    </Button>
                                    <Button variant="primary" onClick={() => handleEditReview(selectedReviewId, editedText)}>
                                        Save Changes
                                    </Button>
                                </Modal.Footer>
                            </Modal>

                            <Button variant="primary" as={Link} to={`/bookings/${destination.id}`} style={{ marginRight: '10px' }}>
                                Book Now
                            </Button>

                            <Button
                                variant="primary"
                                data-testid="delete-destination-button"
                                onClick={() => handleDelete(destination.id)}
                            >
                                delete destination
                            </Button>
                        </div>
                    </ListGroup.Item>
                </ListGroup>
            )
            }
            <Footer />
        </div >
    );
}
