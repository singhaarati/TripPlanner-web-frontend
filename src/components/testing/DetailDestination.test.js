import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom'; // Import the MemoryRouter
import DetailDestination from '../DetailDestination';
// import destinationService from '../../../src/services/destinationService'

test('renders component without error', () => {
    render(
        <MemoryRouter>
            <DetailDestination />
        </MemoryRouter>
    );
});

test('renders loading message', () => {
    // Render the component
    const { getByText } = render(
        <MemoryRouter>
            <DetailDestination />
        </MemoryRouter>);

    // Check if the loading message is displayed
    const loadingMessage = getByText(/Loading.../i);
    expect(loadingMessage).toBeInTheDocument();
});


test('renders DetailDestination correctly', () => {
    // Render the component
    const { asFragment } = render(
        <MemoryRouter>
            <DetailDestination />
        </MemoryRouter>
    );

    // Take a snapshot of the rendered output
    expect(asFragment()).toMatchSnapshot();
});




// test('renders destination name correctly', () => {
//     // Define a sample destination object with a name
//     const destination = {
//       id: 1,
//       name: 'Sample Destination',
//       location: 'Sample Location',
//       price: '$100',
//       picture: 'sample.jpg',
//     };

//     // Render the component, passing the sample destination as a prop
//     const { getByText } = render(
//       <MemoryRouter>
//         <DetailDestination destination={destination} />
//       </MemoryRouter>
//     );

//     // Check if the destination name is present in the rendered component
//     const destinationNameElement = getByText('Sample Destination');
//     expect(destinationNameElement).toBeInTheDocument();
//   });


// test('delete destination button works correctly', () => {
//     // Mock the useNavigate function from react-router-dom
//     const mockNavigate = jest.fn();
//     jest.mock('react-router-dom', () => ({
//         ...jest.requireActual('react-router-dom'),
//         useNavigate: () => mockNavigate,
//     }));

//     // Render the component
//     render(
//         <MemoryRouter>
//             <DetailDestination />
//         </MemoryRouter>
//     );

//     // Simulate the button click
//     const deleteButton = screen.getByText('delete destination');
//     fireEvent.click(deleteButton);

//     // Verify that useNavigate was called and the destination is deleted
//     expect(mockNavigate).toHaveBeenCalledWith('/');
// });