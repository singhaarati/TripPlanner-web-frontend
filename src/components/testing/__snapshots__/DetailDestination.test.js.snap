// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DetailDestination correctly 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="mb-3 navbar navbar-expand-lg navbar-dark bg-dark"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          <img
            alt="Your Logo"
            class="d-inline-block align-top"
            height="30"
            src="logo.png"
            style="width: 50px; height: auto; border: 2px solid #fff; border-radius: 50%; filter: grayscale(50%); transition: filter 0.3s ease;"
          />
        </a>
        <button
          aria-controls="navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="navbar-nav"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              aria-current="page"
              class="nav-link active"
              href="/"
            >
              Home
            </a>
            <a
              class="nav-link"
              href="/aboutus"
            >
              About Us
            </a>
            <a
              class="nav-link"
              href="/MyBookings"
            >
              My Bookings
            </a>
            <a
              class="nav-link"
              href="/AddDestination"
            >
              Add Destination
            </a>
          </div>
          <form
            class="d-flex"
          >
            <input
              aria-label="Search"
              class="me-2 form-control"
              placeholder="Search"
              type="search"
            />
          </form>
          <div
            class="dropdown"
          >
            <button
              aria-expanded="false"
              class="dropdown-toggle btn btn-success"
              data-testid="dropdown-toggle"
              id="dropdown-basic"
              type="button"
            >
              Welcome
            </button>
          </div>
        </div>
      </div>
    </nav>
    <div>
      Loading...
    </div>
    <footer
      data-testid="footer"
      style="background-color: rgb(51, 51, 51); color: rgb(255, 255, 255); padding: 30px 0px;"
    >
      <div
        class="container"
      >
        <div
          style="display: flex; flex-wrap: wrap; justify-content: space-between;"
        >
          <div
            style="flex: 1; margin: 0px 10px;"
          >
            <h3
              style="font-size: 1.5rem; margin-bottom: 1rem;"
            >
              About Us
            </h3>
            <p
              style="margin-bottom: 0.5rem;"
            >
              The Trip Planner app simplifies travel planning,allowing users to overview different destinations, and book destination they desired.
            </p>
          </div>
          <div
            style="flex: 1; margin: 0px 10px;"
          >
            <h3
              style="font-size: 1.5rem; margin-bottom: 1rem;"
            >
              Contact
            </h3>
            <p
              style="margin-bottom: 0.5rem;"
            >
              Thamel, Kathmandu, Nepal
            </p>
            <p
              style="margin-bottom: 0.5rem;"
            >
              Email: tripPlanner@company.com
            </p>
            <p
              style="margin-bottom: 0.5rem;"
            >
              Phone: +977 9836283904
            </p>
          </div>
          <div
            style="flex: 1; margin: 0px 10px;"
          >
            <h3
              style="font-size: 1.5rem; margin-bottom: 1rem;"
            >
              Follow Us
            </h3>
            <div>
              <a
                href="https://www.facebook.com"
                rel="noopener noreferrer"
                target="_blank"
                title="Facebook"
              >
                <i
                  class="fab fa-facebook-f"
                />
              </a>
              <a
                href="https://www.twitter.com/yourcompany"
                rel="noopener noreferrer"
                target="_blank"
                title="Twitter"
              >
                <i
                  class="fab fa-twitter"
                />
              </a>
              <a
                href="https://www.instagram.com/yourcompany"
                rel="noopener noreferrer"
                target="_blank"
                title="Instagram"
              >
                <i
                  class="fab fa-instagram"
                />
              </a>
            </div>
          </div>
        </div>
        <div
          class="text-center"
          style="margin-top: 20px;"
        >
          <span>
            Â© 2023 Trip Planner. All rights reserved.
          </span>
        </div>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
